docs/antrea-l7-network-policy.md

```diff
@@ -0,0 +1,119 @@
+
+## Prerequisites
+
+Layer 7 NetworkPolicy was introduced in v1.10 as an alpha feature and is disabled by default. A feature gate,
+`L7NetworkPolicy` must be enabled in antrea-controller.conf and antrea-agent.conf in the `antrea-config` ConfigMap.
+Besides, due to the constraint of the application detection engine, TX checksum offloading must be disabled via the
+`disableTXChecksumOffload` option in antrea-agent.conf for the feature to work. An example configuration is as below:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: antrea-config
+  namespace: kube-system
+data:
+  antrea-agent.conf: |
+    disableTXChecksumOffload: true
+    featureGates:
+      L7NetworkPolicy: true
+  antrea-controller.conf: |
+    featureGates:
+      L7NetworkPolicy: true
+```
+
+Alternatively, you can use the following helm installation command to configure the above options:
+
+```bash
+helm install antrea antrea/antrea --namespace kube-system --set featureGates.L7NetworkPolicy=true,disableTXChecksumOffload=true
+```
+
+## Usages
+
+There isn't a separate resource type for layer 7 NetworkPolicy. It is one kind of Antrea-native policies which have the
+`l7Protocols` field specified in the rules. Like layer 3 and layer 4 policies, the `l7Protocols` field can be specified
+for ingress and egress rules in Antrea ClusterNetworkPolicy and Antrea NetworkPolicy. It can be used with the `from` or
+`to` fields to select the network peer, and the `ports` to select the transport protocol and/or port for which the layer
+7 rule applies to. The `action` of a layer 7 rule can only be `Allow`.
+
+**Note**: Any traffic matching the layer 3/4 criteria (specified by `from`, `to`, and `port`) of a layer 7 rule will be
+forwarded to an application-aware engine for protocol detection and rule enforcement, and the traffic will be allowed if
+the layer 7 criteria is also matched, otherwise it will be dropped. Therefore, any rules after a layer 7 rule will not
+be enforced for the traffic that match the layer 7 rule's layer 3/4 criteria.
+
+As of now, the only supported layer 7 protocol is HTTP. More protocols will be supported in the near future, and opening
+feature requests for which protocol you are interested in is welcome.
+
+### HTTP
+
+A typical layer 7 NetworkPolicy for HTTP protocol is as below:
+
+```yaml
+apiVersion: crd.antrea.io/v1alpha1
+kind: NetworkPolicy
+metadata:
+  name: ingress-allow-http-request-to-api-v2
+spec:
+  priority: 5
+  tier: application
+  appliedTo:
+    - podSelector:
+        matchLabels:
+          app: web
+  ingress:
+    - name: allow-http   # Allow inbound HTTP GET requests to "/api/v2" from Pods with app=client label
+      action: Allow
+      from:
+        - podSelector:
+            matchLabels:
+              app: client
+      l7Protocols:
+        - http:
+            path: "/api/v2"
+            host: "foo.bar.com"
+            method: "GET"
+    - name: drop-other   # Drop other inbound traffic
```

Comment:
Related to above review, since there had been a discussion on implicit isolation for `from/to` not `appliedTo`, detailed comments would help, and shall we note this somewhere?

Reply:
Thanks, that part looks clear to me.